plonkish_backend/src/pcs/multilinear/mercury.rs

```diff
@@ -0,0 +1,452 @@
+            Evaluation::new(0, 1, g.evaluate(&zeta_inv)), // g(1/ζ)
+            Evaluation::new(1, 0, h.evaluate(&zeta)),     // h(ζ)
+            Evaluation::new(1, 1, h.evaluate(&zeta_inv)), // h(1/ζ)
+            Evaluation::new(2, 0, s.evaluate(&zeta)),     // s(ζ)
+            Evaluation::new(2, 1, s.evaluate(&zeta_inv)), // s(1/ζ)
+            Evaluation::new(3, 0, d.evaluate(&zeta)),     // d(ζ)
+            Evaluation::new(1, 2, h.evaluate(&alpha)),    // h(α)
+            Evaluation::new(4, 0, g_zeta),                // φ(ζ)
+        ];
+        transcript.write_field_elements(evals[..6].iter().map(Evaluation::value))?;
+
+        UnivariateKzg::batch_open(pp, polys, comms, &points, &evals, transcript)
+    }
+
+    fn batch_open<'a>(
+        pp: &Self::ProverParam,
+        polys: impl IntoIterator<Item = &'a Self::Polynomial>,
+        comms: impl IntoIterator<Item = &'a Self::Commitment>,
+        points: &[Point<M::Scalar, Self::Polynomial>],
+        evals: &[Evaluation<M::Scalar>],
+        transcript: &mut impl TranscriptWrite<Self::CommitmentChunk, M::Scalar>,
+    ) -> Result<(), Error>
+    where
+        Self::Commitment: 'a,
+    {
+        let polys = polys.into_iter().collect_vec();
+        let comms = comms.into_iter().collect_vec();
+        let num_vars = points.first().map(|point| point.len()).unwrap_or_default();
+        additive::batch_open::<_, Self>(pp, num_vars, polys, comms, points, evals, transcript)
+    }
+
+    fn read_commitments(
+        vp: &Self::VerifierParam,
+        num_polys: usize,
+        transcript: &mut impl TranscriptRead<Self::CommitmentChunk, M::Scalar>,
+    ) -> Result<Vec<Self::Commitment>, Error> {
+        UnivariateKzg::read_commitments(vp, num_polys, transcript)
+    }
+
+    fn verify(
+        vp: &Self::VerifierParam,
+        comm: &Self::Commitment,
+        point: &Point<M::Scalar, Self::Polynomial>,
+        eval: &M::Scalar,
+        transcript: &mut impl TranscriptRead<Self::CommitmentChunk, M::Scalar>,
+    ) -> Result<(), Error> {
+        let num_vars = point.len();
+        let t = (num_vars >> 1) + 1;
+
+        let h_comm = transcript.read_commitment()?;
+
+        let alpha = transcript.squeeze_challenge();
+        let [g_comm, q_comm] = transcript.read_commitments(2)?.try_into().unwrap();
+
+        let gamma = transcript.squeeze_challenge();
+        let [s_comm, d_comm] = transcript.read_commitments(2)?.try_into().unwrap();
+
+        let zeta = transcript.squeeze_challenge();
+        let zeta_inv = zeta.invert().unwrap();
+
+        // g(ζ), g(1/ζ), h(ζ), h(1/ζ), s(ζ), s(1/ζ)
+        let evals = transcript.read_field_elements(6)?;
+
+        let zeta_squares = squares(zeta).take(t).collect_vec();
+        let zeta_inv_squares = squares(zeta_inv).take(t).collect_vec();
+        let zeta_pow_to_b = zeta_squares
+            .last()
+            .copied()
+            .map(|zeta_pow| zeta_pow.square())
+            .unwrap();
```

Comment:
Could be simplified a bit.

```suggestion
        let zeta_pow_to_b = zeta_squares.last().unwrap().square();
```